### PR TITLE
fs: Re-implement object layer to remember the fd.

### PR DESCRIPTION
**Pull request by harshavardhana**
## Description
This patch re-writes FS backend to support shared backend sharing locks for safe concurrent access across multiple servers.

## Motivation and Context
To facilitate a way to run many minio servers on the same shared backend.

## How Has This Been Tested?
Manually using special setup with shared backend, NFS, XFS and SMB (Windows). 

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: cmd/fs-v1-multipart.go [hunk 1]
```diff
@@ -601,52 +755,42 @@ func (fs fsObjects) CompleteMultipartUpload(bucket string, object string, upload
...
+	if err != nil {
+		return "", toObjectErr(traceError(err), bucket, object)
 	}
+	defer rwlk.Close()
 
-	// Return md5sum.
-	return s3MD5, nil
-}
+	// Close lock held on bucket/object/uploadid/fs.json
+	fs.rwPool.Close(fsMetaPathMultipart)
```
Comment (krishnasrinivas) [inline cmd/fs-v1-multipart.go:764]: this is already done at line 642 and hence doing it again can cause undesired behavior

File: cmd/fs-v1-multipart.go [hunk 2]
```diff
@@ -666,17 +810,52 @@ func (fs fsObjects) AbortMultipartUpload(bucket, object, uploadID string) error
...
-		return traceError(InvalidUploadID{UploadID: uploadID})
+	fsMetaPath := pathJoin(fs.fsPath, minioMetaMultipartBucket, uploadIDPath, fsMetaJSONFile)
+	_, err = fs.rwPool.Open(fsMetaPath)
+	if err != nil {
+		if err == errFileNotFound {
+			return traceError(InvalidUploadID{UploadID: uploadID})
+		}
+		return toObjectErr(traceError(err), bucket, object)
 	}
+	fs.rwPool.Close(fsMetaPath)
```
Comment (krishnasrinivas) [inline cmd/fs-v1-multipart.go]: should do `defer fs.rwPool.Close(fsMetaPath)`

File: cmd/fs-v1-rwpool.go [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+// count has reached zero.
+func (fsi *fsIOPool) Open(path string) (*lock.RLockedFile, error) {
+	fsi.Lock()
+	rlkFile, ok := fsi.readersMap[path]
+
+	// File reference exists on map, validate if its
+	// really closed and we are safe to purge it.
+	if ok {
+		// If the file is closed, remove it from the reader pool map.
+		if rlkFile.IsClosed() {
```
Comment (krishnasrinivas) [inline cmd/fs-v1-rwpool.go:58]: should never reach here, we can do an errorIf() to catch a bug

File: cmd/fs-v1-multipart.go (hunk 1 above)
Comment (harshavardhana) [reply cmd/fs-v1-multipart.go:764]: Needed on windows, its not really undesired.

File: cmd/fs-v1-multipart.go (hunk 2 above)
Comment (harshavardhana) [reply cmd/fs-v1-multipart.go]: Same windows its needed.

File: cmd/fs-v1-rwpool.go (hunk 3 above)
Comment (harshavardhana) [reply cmd/fs-v1-rwpool.go:58]: Done

File: cmd/fs-v1-multipart.go (hunk 1 above)
Comment (krishnasrinivas) [reply cmd/fs-v1-multipart.go:764]: double Close() will cause problems, i.e it can close the fd even when its not supposed to.

Comment (harshavardhana) [reply cmd/fs-v1-multipart.go:764]: done

File: cmd/fs-v1-multipart.go (hunk 2 above)
Comment (harshavardhana) [reply cmd/fs-v1-multipart.go]: done

**Pull request merged**